File: src/main/java/net/jgp/books/spark/ch13/lab110_flatten_shipment/FlattenShipmentDisplayApp.java
package net.jgp.books.spark.ch13.lab110_flatten_shipment;

import static org.apache.spark.sql.functions.explode;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * Processing of invoices formatted using the schema.org format.
 * 
 * @author jgp
 */
public class FlattenShipmentDisplayApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    FlattenShipmentDisplayApp app = new FlattenShipmentDisplayApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Flatenning JSON doc describing shipments")
        .master("local")
        .getOrCreate();

    // Reads a JSON, stores it in a dataframe
    Dataset<Row> df = spark.read()
        .format("json")
        .option("multiline", true)
        .load("data/json/shipment.json");

    df = df
        .withColumn("supplier_name", df.col("supplier.name"))
        .withColumn("supplier_city", df.col("supplier.city"))
        .withColumn("supplier_state", df.col("supplier.state"))
        .withColumn("supplier_country", df.col("supplier.country"))
        .drop("supplier")
        .withColumn("customer_name", df.col("customer.name"))
        .withColumn("customer_city", df.col("customer.city"))
        .withColumn("customer_state", df.col("customer.state"))
        .withColumn("customer_country", df.col("customer.country"))
        .drop("customer")
        .withColumn("items", explode(df.col("books")));
    df = df
        .withColumn("qty", df.col("items.qty"))
        .withColumn("title", df.col("items.title"))
        .drop("items")
        .drop("books");

    // Shows at most 5 rows from the dataframe (there's only one anyway)
    df.show(5, false);
    df.printSchema();

    df.createOrReplaceTempView("shipment_detail");
    Dataset<Row> bookCountDf =
        spark.sql("SELECT COUNT(*) AS bookCount FROM shipment_detail");
    bookCountDf.show(false);
  }
}

File: src/main/python/lab110_flatten_shipment/flattenShipmentDisplayApp.py
"""
 Processing of invoices formatted using the schema.org format.

 @author rambabu.posa
"""
import os
from pyspark.sql import (SparkSession, functions as F)

def get_absolute_file_path(path, filename):
    # To get absolute path for a given filename
    current_dir = os.path.dirname(__file__)
    relative_path = "{}{}".format(path, filename)
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path

def main(spark):
    # The processing code.
    filename = 'shipment.json'
    path = '../../../../data/json/'
    absolute_file_path = get_absolute_file_path(path, filename)

    # Reads a JSON, stores it in a dataframe
    # Dataset[Row] == DataFrame
    df = spark.read.format("json") \
        .option("multiline", True) \
        .load(absolute_file_path)

    df = df \
        .withColumn("supplier_name", F.col("supplier.name")) \
        .withColumn("supplier_city", F.col("supplier.city")) \
        .withColumn("supplier_state", F.col("supplier.state")) \
        .withColumn("supplier_country", F.col("supplier.country")) \
        .drop("supplier") \
        .withColumn("customer_name", F.col("customer.name")) \
        .withColumn("customer_city", F.col("customer.city")) \
        .withColumn("customer_state", F.col("customer.state")) \
        .withColumn("customer_country", F.col("customer.country")) \
        .drop("customer") \
        .withColumn("items", F.explode(F.col("books")))

    df = df \
        .withColumn("qty", F.col("items.qty")) \
        .withColumn("title", F.col("items.title")) \
        .drop("items") \
        .drop("books")

    # Shows at most 5 rows from the dataframe (there's only one anyway)
    df.show(5, False)
    df.printSchema()

    df.createOrReplaceTempView("shipment_detail")

    sqlQuery = "SELECT COUNT(*) AS bookCount FROM shipment_detail"
    bookCountDf = spark.sql(sqlQuery)

    bookCountDf.show()

if __name__ == '__main__':
    # Creates a session on a local master
    spark = SparkSession.builder.appName("Flatenning JSON doc describing shipments") \
    .master("local[*]").getOrCreate()

    # Comment this line to see full log
    spark.sparkContext.setLogLevel('error')
    main(spark)
    spark.stop()
